File: source/tree/splaytree/c/splay_tree.c
/**
 * SplayTree伸展树(C语言): C语言实现的伸展树。
 *
 * @author skywang
 * @date 2014/02/03
 */

#include <stdio.h>
#include <stdlib.h>
#include "splay_tree.h"

/*
 * 前序遍历"伸展树"
 */
void preorder_splaytree(SplayTree tree)
{
	if(tree != NULL)
	{
		printf("%d ", tree->key);
		preorder_splaytree(tree->left);
		preorder_splaytree(tree->right);
	}
}

/*
 * 中序遍历"伸展树"
 */
void inorder_splaytree(SplayTree tree)
{
	if(tree != NULL)
	{
		inorder_splaytree(tree->left);
		printf("%d ", tree->key);
		inorder_splaytree(tree->right);
	}
}

/*
 * 后序遍历"伸展树"
 */
void postorder_splaytree(SplayTree tree)
{
	if(tree != NULL)
	{
		postorder_splaytree(tree->left);
		postorder_splaytree(tree->right);
		printf("%d ", tree->key);
	}
}

/*
 * (递归实现)查找"伸展树x"中键值为key的节点
 */
Node* splaytree_search(SplayTree x, Type key)
{
	if (x==NULL || x->key==key)
		return x;

	if (key < x->key)
		return splaytree_search(x->left, key);
	else
		return splaytree_search(x->right, key);
}

/*
 * (非递归实现)查找"伸展树x"中键值为key的节点
 */
Node* iterative_splaytree_search(SplayTree x, Type key)
{
	while ((x!=NULL) && (x->key!=key))
	{
		if (key < x->key)
			x = x->left;
		else
			x = x->right;
	}

	return x;
}

/* 
 * 查找最小结点：返回tree为根结点的伸展树的最小结点。
 */
Node* splaytree_minimum(SplayTree tree)
{
	if (tree == NULL)
		return NULL;

	while(tree->left != NULL)
		tree = tree->left;
	return tree;
}
 
/* 
 * 查找最大结点：返回tree为根结点的伸展树的最大结点。
 */
Node* splaytree_maximum(SplayTree tree)
{
	if (tree == NULL)
		return NULL;

	while(tree->right != NULL)
		tree = tree->right;
	return tree;
}

/* 
 * 旋转key对应的节点为根节点，并返回根节点。
 *
 * 注意：
 *   (a)：伸展树中存在"键值为key的节点"。
 *          将"键值为key的节点"旋转为根节点。
 *   (b)：伸展树中不存在"键值为key的节点"，并且key < tree->key。
 *      b-1 "键值为key的节点"的前驱节点存在的话，将"键值为key的节点"的前驱节点旋转为根节点。
 *      b-2 "键值为key的节点"的前驱节点存在的话，则意味着，key比树中任何键值都小，那么此时，将最小节点旋转为根节点。
 *   (c)：伸展树中不存在"键值为key的节点"，并且key > tree->key。
 *      c-1 "键值为key的节点"的后继节点存在的话，将"键值为key的节点"的后继节点旋转为根节点。
 *      c-2 "键值为key的节点"的后继节点不存在的话，则意味着，key比树中任何键值都大，那么此时，将最大节点旋转为根节点。
 */
Node* splaytree_splay(SplayTree tree, Type key)
{
    Node N, *l, *r, *c;

    if (tree == NULL) 
        return tree;

    N.left = N.right = NULL;
    l = r = &N;

    for (;;)
    {
        if (key < tree->key)
        {
            if (tree->left == NULL)
                break;
            if (key < tree->left->key)
            {
                c = tree->left;                           /* 01, rotate right */
                tree->left = c->right;
                c->right = tree;
                tree = c;
                if (tree->left == NULL) 
                    break;
            }
            r->left = tree;                               /* 02, link right */
            r = tree;
            tree = tree->left;
        }
		else if (key > tree->key)
        {
            if (tree->right == NULL) 
                break;
            if (key > tree->right->key) 
            {
                c = tree->right;                          /* 03, rotate left */
                tree->right = c->left;
                c->left = tree;
                tree = c;
                if (tree->right == NULL) 
                    break;
            }
            l->right = tree;                              /* 04, link left */
            l = tree;
            tree = tree->right;
        }
		else
		{
            break;
        }
    }

    l->right = tree->left;                                /* 05, assemble */
    r->left = tree->right;
    tree->left = N.right;
    tree->right = N.left;

    return tree;
}

/* 
 * 将结点插入到伸展树中(不旋转)
 *
 * 参数说明：
 *     tree 伸展树的根结点
 *     z 插入的结点
 * 返回值：
 *     根节点
 */
static Node* splaytree_insert(SplayTree tree, Node *z)
{
	Node *y = NULL;
	Node *x = tree;

	// 查找z的插入位置
	while (x != NULL)
	{
		y = x;
		if (z->key < x->key)
			x = x->left;
		else if (z->key > x->key)
			x = x->right;
		else
		{
			printf("不允许插入相同节点(%d)!\n", z->key);
			// 释放申请的节点，并返回tree。
			free(z);
			return tree;
		}
	}

	if (y==NULL)
		tree = z;
	else if (z->key < y->key)
		y->left = z;
	else
		y->right = z;

	return tree;
}

/*
 * 创建并返回伸展树结点。
 *
 * 参数说明：
 *     key 是键值。
 *     parent 是父结点。
 *     left 是左孩子。
 *     right 是右孩子。
 */
static Node* create_splaytree_node(Type key, Node *left, Node* right)
{
	Node* p;

	if ((p = (Node *)malloc(sizeof(Node))) == NULL)
		return NULL;
	p->key = key;
	p->left = left;
	p->right = right;

	return p;
}

/* 
 * 新建结点(key)，然后将其插入到伸展树中，并将插入节点旋转为根节点
 *
 * 参数说明：
 *     tree 伸展树的根结点
 *     key 插入结点的键值
 * 返回值：
 *     根节点
 */
Node* insert_splaytree(SplayTree tree, Type key)
{
	Node *z;	// 新建结点

	// 如果新建结点失败，则返回。
	if ((z=create_splaytree_node(key, NULL, NULL)) == NULL)
		return tree;

	// 插入节点
	tree = splaytree_insert(tree, z);
	// 将节点(key)旋转为根节点
	tree = splaytree_splay(tree, key);
}

/* 
 * 删除结点(key为节点的键值)，并返回根节点。
 *
 * 参数说明：
 *     tree 伸展树的根结点
 *     z 删除的结点
 * 返回值：
 *     根节点(根节点是被删除节点的前驱节点)
 *
 */
Node* delete_splaytree(SplayTree tree, Type key)
{
    Node *x;

    if (tree == NULL) 
        return NULL;

	// 查找键值为key的节点，找不到的话直接返回。
	if (splaytree_search(tree, key) == NULL)
		return tree;

	// 将key对应的节点旋转为根节点。
    tree = splaytree_splay(tree, key);

	if (tree->left != NULL)
	{
		// 将"tree的前驱节点"旋转为根节点
		x = splaytree_splay(tree->left, key);
		// 移除tree节点
		x->right = tree->right;
	}
	else
		x = tree->right;

	free(tree);

	return x;
}

/*
 * 销毁伸展树
 */
void destroy_splaytree(SplayTree tree)
{
	if (tree==NULL)
		return ;

	if (tree->left != NULL)
		destroy_splaytree(tree->left);
	if (tree->right != NULL)
		destroy_splaytree(tree->right);

	free(tree);
}

/*
 * 打印"伸展树"
 *
 * tree       -- 伸展树的节点
 * key        -- 节点的键值 
 * direction  --  0，表示该节点是根节点;
 *               -1，表示该节点是它的父结点的左孩子;
 *                1，表示该节点是它的父结点的右孩子。
 */
void print_splaytree(SplayTree tree, Type key, int direction)
{
	if(tree != NULL)
	{
		if(direction==0)	// tree是根节点
			printf("%2d is root\n", tree->key);
		else				// tree是分支节点
			printf("%2d is %2d's %6s child\n", tree->key, key, direction==1?"right" : "left");

		print_splaytree(tree->left, tree->key, -1);
		print_splaytree(tree->right,tree->key,  1);
	}
}

File: source/tree/splaytree/java/SplayTree.java
/**
 * Java 语言: 伸展树
 *
 * @author skywang
 * @date 2014/02/03
 */

public class SplayTree<T extends Comparable<T>> {

	private SplayTreeNode<T> mRoot;	// 根结点

	public class SplayTreeNode<T extends Comparable<T>> {
		T key;				// 关键字(键值)
		SplayTreeNode<T> left;	// 左孩子
		SplayTreeNode<T> right;	// 右孩子

		public SplayTreeNode() {
			this.left = null;
			this.right = null;
		}

		public SplayTreeNode(T key, SplayTreeNode<T> left, SplayTreeNode<T> right) {
			this.key = key;
			this.left = left;
			this.right = right;
		}
	}

	public SplayTree() {
		mRoot=null;
	}

	/*
	 * 前序遍历"伸展树"
	 */
	private void preOrder(SplayTreeNode<T> tree) {
		if(tree != null) {
			System.out.print(tree.key+" ");
			preOrder(tree.left);
			preOrder(tree.right);
		}
	}

	public void preOrder() {
		preOrder(mRoot);
	}

	/*
	 * 中序遍历"伸展树"
	 */
	private void inOrder(SplayTreeNode<T> tree) {
		if(tree != null) {
			inOrder(tree.left);
			System.out.print(tree.key+" ");
			inOrder(tree.right);
		}
	}

	public void inOrder() {
		inOrder(mRoot);
	}


	/*
	 * 后序遍历"伸展树"
	 */
	private void postOrder(SplayTreeNode<T> tree) {
		if(tree != null)
		{
			postOrder(tree.left);
			postOrder(tree.right);
			System.out.print(tree.key+" ");
		}
	}

	public void postOrder() {
		postOrder(mRoot);
	}


	/*
	 * (递归实现)查找"伸展树x"中键值为key的节点
	 */
	private SplayTreeNode<T> search(SplayTreeNode<T> x, T key) {
		if (x==null)
			return x;

		int cmp = key.compareTo(x.key);
		if (cmp < 0)
			return search(x.left, key);
		else if (cmp > 0)
			return search(x.right, key);
		else
			return x;
	}

	public SplayTreeNode<T> search(T key) {
		return search(mRoot, key);
	}

	/*
	 * (非递归实现)查找"伸展树x"中键值为key的节点
	 */
	private SplayTreeNode<T> iterativeSearch(SplayTreeNode<T> x, T key) {
		while (x!=null) {
			int cmp = key.compareTo(x.key);

			if (cmp < 0) 
				x = x.left;
			else if (cmp > 0) 
				x = x.right;
			else
				return x;
		}

		return x;
	}

	public SplayTreeNode<T> iterativeSearch(T key) {
		return iterativeSearch(mRoot, key);
	}

	/* 
	 * 查找最小结点：返回tree为根结点的伸展树的最小结点。
	 */
	private SplayTreeNode<T> minimum(SplayTreeNode<T> tree) {
		if (tree == null)
			return null;

		while(tree.left != null)
			tree = tree.left;
		return tree;
	}

	public T minimum() {
		SplayTreeNode<T> p = minimum(mRoot);
		if (p != null)
			return p.key;

		return null;
	}
	 
	/* 
	 * 查找最大结点：返回tree为根结点的伸展树的最大结点。
	 */
	private SplayTreeNode<T> maximum(SplayTreeNode<T> tree) {
		if (tree == null)
			return null;

		while(tree.right != null)
			tree = tree.right;
		return tree;
	}

	public T maximum() {
		SplayTreeNode<T> p = maximum(mRoot);
		if (p != null)
			return p.key;

		return null;
	}

	/* 
	 * 旋转key对应的节点为根节点，并返回根节点。
	 *
	 * 注意：
	 *   (a)：伸展树中存在"键值为key的节点"。
	 *          将"键值为key的节点"旋转为根节点。
	 *   (b)：伸展树中不存在"键值为key的节点"，并且key < tree.key。
	 *      b-1 "键值为key的节点"的前驱节点存在的话，将"键值为key的节点"的前驱节点旋转为根节点。
	 *      b-2 "键值为key的节点"的前驱节点存在的话，则意味着，key比树中任何键值都小，那么此时，将最小节点旋转为根节点。
	 *   (c)：伸展树中不存在"键值为key的节点"，并且key > tree.key。
	 *      c-1 "键值为key的节点"的后继节点存在的话，将"键值为key的节点"的后继节点旋转为根节点。
	 *      c-2 "键值为key的节点"的后继节点不存在的话，则意味着，key比树中任何键值都大，那么此时，将最大节点旋转为根节点。
	 */
	private SplayTreeNode<T> splay(SplayTreeNode<T> tree, T key) {
		if (tree == null) 
			return tree;

		SplayTreeNode<T> N = new SplayTreeNode<T>();
		SplayTreeNode<T> l = N;
		SplayTreeNode<T> r = N;
		SplayTreeNode<T> c;

		for (;;) {

			int cmp = key.compareTo(tree.key);
			if (cmp < 0) {

				if (tree.left == null)
					break;

				if (key.compareTo(tree.left.key) < 0) {
					c = tree.left;                           /* rotate right */
					tree.left = c.right;
					c.right = tree;
					tree = c;
					if (tree.left == null) 
						break;
				}
				r.left = tree;                               /* link right */
				r = tree;
				tree = tree.left;
			} else if (cmp > 0) {

				if (tree.right == null) 
					break;

				if (key.compareTo(tree.right.key) > 0) {
					c = tree.right;                          /* rotate left */
					tree.right = c.left;
					c.left = tree;
					tree = c;
					if (tree.right == null) 
						break;
				}

				l.right = tree;                              /* link left */
				l = tree;
				tree = tree.right;
			} else {
				break;
			}
		}

		l.right = tree.left;                                /* assemble */
		r.left = tree.right;
		tree.left = N.right;
		tree.right = N.left;

		return tree;
	}

	public void splay(T key) {
		mRoot = splay(mRoot, key);
	}

	/* 
     * 将结点插入到伸展树中，并返回根节点
	 *
	 * 参数说明：
	 *     tree 伸展树的
	 *     z 插入的结点
	 */
	private SplayTreeNode<T> insert(SplayTreeNode<T> tree, SplayTreeNode<T> z) {
		int cmp;
		SplayTreeNode<T> y = null;
		SplayTreeNode<T> x = tree;

		// 查找z的插入位置
		while (x != null) {
			y = x;
			cmp = z.key.compareTo(x.key);
			if (cmp < 0)
				x = x.left;
			else if (cmp > 0)
				x = x.right;
			else {
				System.out.printf("不允许插入相同节点(%d)!\n", z.key);
				z=null;
				return tree;
			}
		}

		if (y==null)
			tree = z;
		else {
			cmp = z.key.compareTo(y.key);
			if (cmp < 0)
				y.left = z;
			else
				y.right = z;
		}

		return tree;
	}

	public void insert(T key) {
		SplayTreeNode<T> z=new SplayTreeNode<T>(key,null,null);

		// 如果新建结点失败，则返回。
		if ((z=new SplayTreeNode<T>(key,null,null)) == null)
			return ;

		// 插入节点
		mRoot = insert(mRoot, z);
		// 将节点(key)旋转为根节点
		mRoot = splay(mRoot, key);
	}

	/* 
	 * 删除结点(z)，并返回被删除的结点
	 *
	 * 参数说明：
	 *     bst 伸展树
	 *     z 删除的结点
	 */
	private SplayTreeNode<T> remove(SplayTreeNode<T> tree, T key) {
		SplayTreeNode<T> x;

		if (tree == null) 
			return null;

		// 查找键值为key的节点，找不到的话直接返回。
		if (search(tree, key) == null)
			return tree;

		// 将key对应的节点旋转为根节点。
		tree = splay(tree, key);

		if (tree.left != null) {
			// 将"tree的前驱节点"旋转为根节点
			x = splay(tree.left, key);
			// 移除tree节点
			x.right = tree.right;
		}
		else
			x = tree.right;

		tree = null;

		return x;
	}

	public void remove(T key) {
		mRoot = remove(mRoot, key);
	}

	/*
	 * 销毁伸展树
	 */
	private void destroy(SplayTreeNode<T> tree) {
		if (tree==null)
			return ;

		if (tree.left != null)
			destroy(tree.left);
		if (tree.right != null)
			destroy(tree.right);

		tree=null;
	}

	public void clear() {
		destroy(mRoot);
		mRoot = null;
	}

	/*
	 * 打印"伸展树"
	 *
	 * key        -- 节点的键值 
	 * direction  --  0，表示该节点是根节点;
	 *               -1，表示该节点是它的父结点的左孩子;
	 *                1，表示该节点是它的父结点的右孩子。
	 */
	private void print(SplayTreeNode<T> tree, T key, int direction) {

		if(tree != null) {

			if(direction==0)	// tree是根节点
				System.out.printf("%2d is root\n", tree.key);
			else				// tree是分支节点
				System.out.printf("%2d is %2d's %6s child\n", tree.key, key, direction==1?"right" : "left");

			print(tree.left, tree.key, -1);
			print(tree.right,tree.key,  1);
		}
	}

	public void print() {
		if (mRoot != null)
			print(mRoot, mRoot.key, 0);
	}
}
